translator/trace/otlp/otlp_to_oc.go

```diff
@@ -0,0 +1,303 @@
+	}
+}
+
+func linksToOCLinks(links []*tracepb.Span_Link) []*octrace.Span_Link {
+	if len(links) == 0 {
+		return nil
+	}
+
+	ocLinks := make([]*octrace.Span_Link, 0, len(links))
+	for _, link := range links {
+		ocLink := &octrace.Span_Link{
+			TraceId:    link.TraceId,
+			SpanId:     link.SpanId,
+			Tracestate: tracestateToOC(link.Tracestate),
+			Attributes: attributesToOCSpanAttributes(link.Attributes, link.DroppedAttributesCount),
+		}
+		ocLinks = append(ocLinks, ocLink)
+	}
+	return ocLinks
+}
+
+func eventsToOC(events []*tracepb.Span_Event, droppedCount uint32) *octrace.Span_TimeEvents {
+	if len(events) == 0 && droppedCount == 0 {
+		return nil
+	}
+
+	return &octrace.Span_TimeEvents{
+		TimeEvent:                 eventsToOCEvents(events),
+		DroppedMessageEventsCount: int32(droppedCount),
+	}
+}
+
+func eventsToOCEvents(events []*tracepb.Span_Event) []*octrace.Span_TimeEvent {
+	if len(events) == 0 {
+		return nil
+	}
+
+	ocEvents := make([]*octrace.Span_TimeEvent, 0, len(events))
+	for _, event := range events {
+		ocEvent := &octrace.Span_TimeEvent{
+			Time: unixnanoToTimestamp(event.TimeUnixnano),
+			Value: &octrace.Span_TimeEvent_Annotation_{
+				Annotation: &octrace.Span_TimeEvent_Annotation{
+					Description: &octrace.TruncatableString{
+						Value: event.Name,
+					},
+					Attributes: attributesToOCSpanAttributes(event.Attributes, event.DroppedAttributesCount),
+				},
+			},
+		}
+		ocEvents = append(ocEvents, ocEvent)
+	}
+	return ocEvents
+}
+
+func statusToOC(status *tracepb.Status) *octrace.Status {
+	if status == nil {
+		return nil
+	}
+	return &octrace.Status{
+		Code:    int32(status.Code),
+		Message: status.Message,
+	}
+}
+
+func resourceToOC(resource *resourcepb.Resource) (*occommon.Node, *ocresource.Resource) {
+	if resource == nil {
+		return nil, nil
+	}
+
+	return &occommon.Node{
+			Attributes: attributesToOC(resource.Attributes),
+		},
+		&ocresource.Resource{
+			Labels: attributesToOC(resource.Attributes),
+		}
+}
+
+func attributesToOC(attributes []*commonpb.AttributeKeyValue) map[string]string {
+	if len(attributes) == 0 {
+		return nil
+	}
+
+	ocAttributes := make(map[string]string, len(attributes))
+	for _, attr := range attributes {
+		ocAttributes[attr.Key] = attributeValueToString(attr)
+	}
+	return ocAttributes
+}
+
+func attributesToOCSpanAttributes(attributes []*commonpb.AttributeKeyValue, droppedCount uint32) *octrace.Span_Attributes {
+	if len(attributes) == 0 && droppedCount == 0 {
+		return nil
+	}
+
+	return &octrace.Span_Attributes{
+		AttributeMap:           attributesToOCMap(attributes),
+		DroppedAttributesCount: int32(droppedCount),
+	}
+}
+
+func attributesToOCMap(attributes []*commonpb.AttributeKeyValue) map[string]*octrace.AttributeValue {
+	if len(attributes) == 0 {
+		return nil
+	}
+
+	ocAttributes := make(map[string]*octrace.AttributeValue, len(attributes))
+	for _, attr := range attributes {
+		ocAttributes[attr.Key] = attributeValueToOC(attr)
```

Comment:
Nit: we could preallocate all `octrace.AttributeValue` structs at once, but this is fine for a temporary code, no need to change anything.